### PR TITLE
Document `Fiber::SchedulerInterface#io_read` and `#io_write`

### PR DESCRIPTION
**Pull request by zverok**
Screenshot (ignore `#io_wait`):
image

### CODE REVIEW

**Comment by zverok** [comment]
@ioquatix Please review. Should other details be mentioned? Is the description correct overall?

**Comment by ioquatix** [comment]
This looks good.

## `io_read`

The length argument is the "minimum length to be read".

If the IO buffer is big, e.g. 8KiB, but the length is 1KiB, up to 8KiB might be read, but at least 1KiB will be.

## `io_write`

The length argument is the "[maximum] length to be written".

If the IO buffer is big, e.g. 8KiB, but the length is 1KiB, at most 1KiB will be written. Generally, the only case where less data can be written is if there is an error writing the data.

**Comment by ioquatix** [comment]
Please mark/tag `io_write` and `io_read` as experimental.

**Comment by ioquatix** [comment]
I had to make some internal changes to the hooks, I will have some updates to this documentation in a day or two.

https://github.com/ruby/ruby/pull/5287

**Comment by zverok** [comment]
@ioquatix Is that what you had in mind? I am not sure—maybe I am lost a bit—whether the new interface is expected to return "length or -errno" (how I described it) or `[-errno, length]` (hmm, that's what your PR's title says?..)

image

**Comment by ioquatix** [comment]
Yes, this interface can return an integer in the range of `[-errno, length]`.

I provided C interface to hide exact implementation details of the `VALUE` returned: https://github.com/ruby/ruby/pull/5287/files#diff-1552659974ee1047f418def875393a583c6a8c543e446755a52acdcae5467fc1R28-R47

FYI Linux internally follows this model and libc writes the return value to `errno`: https://yarchive.net/comp/linux/errno.html from Linus himself.

**Comment by ioquatix** [comment]
I need to revise `io_write` after the above linked PR is merged:

## io_write

The length argument is the "[minimum] length to be written".

If the IO buffer is 8KiB, but the length is 1KiB, up to 8KiB might be written, but at least 1KiB will be.

File: cont.c [hunk 1]
```diff
@@ -3002,6 +3002,61 @@ rb_fiber_scheduler_interface_io_wait(VALUE self)
...
+/*
+ *  Document-method: SchedulerInterface#io_read
+ *  call-seq: io_read(io, buffer, length) -> read length or -errno
+ *
+ *  Invoked by IO#read to read +length+ bytes from +io+ into a specified
+ *  +buffer+ (see IO::Buffer).
+ *
+ *  The +length+ argument is the "minimum length to be read".
+ *  If the IO buffer is big, e.g. 8KiB, but the +length+ is +1024+ (1KiB), up to
+ *  8KiB might be read, but at least 1KiB will be.
```
Comment (ioquatix) [inline cont.c:3014]:
```suggestion
 *  The +length+ argument is the "minimum length to be read".
 *  If the IO buffer size is 8KiB, but the +length+ is +1024+ (1KiB), up to
 *  8KiB might be read, but at least 1KiB will be.
```

File: cont.c [hunk 2]
```diff
@@ -3002,6 +3002,61 @@ rb_fiber_scheduler_interface_io_wait(VALUE self)
...
+ *  Document-method: SchedulerInterface#io_write
+ *  call-seq: io_write(io, buffer, length) -> written length or -errno
+ *
+ *  Invoked by IO#write to write +length+ bytes to +io+ from
+ *  from a specified +buffer+ (see IO::Buffer).
+ *
+ *  The +length+ argument is the "(maximum) length to be written".
+ *  If the IO buffer is big, e.g. 8KiB, but the length is 1024 (1KiB), at most 1KiB
+ *  will be written. Generally, the only case where less data can be written is if
+ *  there is an error writing the data.
```
Comment (ioquatix) [inline cont.c:3042]:
```suggestion
 *  The +length+ argument is the "(maximum) length to be written".
 *  If the IO buffer size is 8KiB, but the length is 1024 (1KiB), at most 8KiB
 *  will be written, but at least 1 KiB will be.
 *  Generally, the only case where less data will be written is if
 *  there is an error writing the data.
 *
 *  Specifying a +length+ of 0 is valid and means write as much as possible.
```

**Comment by ioquatix** [comment]
@zverok thanks so much for helping with this.

**Review by bryanp — COMMENTED** [review]
👍 This is great.

**Comment by zverok** [comment]
@ioquatix I've applied your suggestions, thanks! :hugs: 
I also added two more methods docs (wanted to do in a separate PR, but it actually makes more sense to handle it all once and for good):

**`#address_resolve`**
image

**`#timeout_after`**
image

**Comment by ioquatix** [comment]
I wonder if it would make sense to move this pseudo-implementation to `scheduler.c` just to avoid making `cont.c` more burdened.

File: cont.c [hunk 3]
```diff
@@ -3018,6 +3084,41 @@ rb_fiber_scheduler_interface_kernel_sleep(VALUE self)
...
+ *  call-seq: timeout_after(duration, exception_class, *exception_args, &block) -> obj
+ *
+ *  Invoked by Timeout.timeout to perform execution time control in a non-blocking way.
+ *
+ *  The method is expected to execute a +block+, and if its execution takes longer
+ *  than +duration+, to raise +exception_class+ constructed with +exception_args+.
+ *
+ *  If the block is executed successfully, its result should be returned.
+ *
+ *  The method support should be considered _experimental_.
```
Comment (ioquatix) [inline cont.c:3115]:
```suggestion
 *  Document-method: SchedulerInterface#address_resolve
 *  call-seq: timeout_after(duration, exception_class, *exception_arguments, &block) -> result of block
 *
 *  Limit the execution time of a given +block+ to the given +duration+ if
 *  possible. When a non-blocking operation causes the +block+'s execution time
 *  to exceed the specified +duration+, that non-blocking operation should be
 *  interrupted by raising the specified +exception_class+ constructed with the
 *  given +exception_arguments+.
 *
 *  This hook is invoked by Timeout.timeout and can also be invoked directly on
 *  the scheduler.
 *
 *  If the block is executed successfully, its result will be returned.
 *
 *  The exception will typically be raised using Fiber#raise.
```

I don't know if it's valid to reference it using `Fiber#raise`.

**Comment by ioquatix** [comment]
@zverok I've asked the author of `address_resolve` hook @bruno- to provide/update the documentation.

File: cont.c [hunk 4]
```diff
@@ -3018,6 +3084,41 @@ rb_fiber_scheduler_interface_kernel_sleep(VALUE self)
...
+ */
+static VALUE
+rb_fiber_scheduler_interface_address_resolve(VALUE self)
+{
+}
+
+/*
+ *  Document-method: SchedulerInterface#address_resolve
```
Comment (bruno-) [inline cont.c:3105]: This line should contain `SchedulerInterface#timeout_after`, right?

File: cont.c [hunk 5]
```diff
@@ -3018,6 +3084,41 @@ rb_fiber_scheduler_interface_kernel_sleep(VALUE self)
 {
 }
 
+/*
+ *  Document-method: SchedulerInterface#address_resolve
+ *  call-seq: address_resolve(hostname) -> array_of_stings or nil
```
Comment (bruno-) [inline cont.c:3089]: array_of_stRings

File: cont.c [hunk 6]
```diff
@@ -3018,6 +3084,41 @@ rb_fiber_scheduler_interface_kernel_sleep(VALUE self)
...
+ *
+ *  Invoked by Socket::getaddrinfo and is expected to provide hostname resolution
+ *  in a non-blocking way.
```
Comment (bruno-) [inline cont.c:3092]: This hook is also called by a wide variety of other methods in Ruby. Any (non-reverse) DNS query should invoke this hook. Here's a doc proposal:

```
Invoked by any method that performs a non-reverse DNS lookup. The most notable method is Addrinfo::getaddrinfo, but there are many other, like Socket::getaddrinfo, IPSocket.getaddress, TCPServer.new etc.
```

Here's the full list of methods: https://gist.github.com/bruno-/d2cf7690b0aa4bfa03a8f6823102ffc8

File: cont.c (hunk 4 above)
Comment (ioquatix) [reply cont.c:3105]: Thanks, fixed.

File: cont.c (hunk 5 above)
Comment (ioquatix) [reply cont.c:3089]: Thanks, fixed.

File: cont.c (hunk 6 above)
Comment (ioquatix) [reply cont.c:3092]: Thanks, added.

**Pull request merged**